enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,458 @@
+- Automatic sorting of available disks into StorageClasses based on their characteristics.
+- Detect newly attached disks
+
+### Non-Goals
+
+- Inherently protect against confilict with provisioners that own local devices on the same nodes automatic detection is configured to run.
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+
+## Proposal
+
+The `local-storage-operator` is already capable of consuming local disks from the nodes and provisioning PVs out of them,
+but the disk/device paths needs to explicitly specified in the `LocalVolume` CR.
+The idea is to introduce two new features in LSO
+  - Auto discovery of local devices
+    - This will involve the intoduction of two new CRs `LocalVolumeDiscovery` and `LocalVolumeNodeDiscovery`.
+    - The purpose of this will be to expose local devices available in a node via the `LocalVolumeNodeDiscovery` CR.
+    - The device discovery will be done once via a discovery job.
+  - Auto discovery and provisioning of localDevices
+    - This will involve introduction of a new CR called `AutoDetectVolume`.
+    -  The pupose of this will be to auto discover and provision PVs on devices which match the inclusion filter present in the CR.
+    -  This will be a continous process.
+
+Once we have the detected devices the administrator can create the localPVs by explicitly selecting the disks. This can be done by the `localVolume` CR. The other option would be to create localPVs via the AutoDetectVolume CR and passing the inclusion filters in it.
+
+
+## Design Details for `Auto discovery of local devices`
+
+API scheme for `LocalVolumeDiscovery` and `LocalVolumeNodeDiscovery` CR:
+
+```go
+type LocalVolumeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscovery `json:"items"`
+}
+
+type LocalVolumeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoverySpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoverySpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type LocalVolumeDiscoveryStatus struct {
+	Phase string `json:"phase,omitempty"`
```

Comment:
What are the possible phases?

Reply:
I have documented it: 
```
Possible status.phase for LocalVolumeDiscovery CR:
- Deploying
- Failed
- Deployed
```

Reply:
I have documented it
```Possible `status.phase` for `LocalVolumeDiscovery` CR:
    - `InProgress`: Generating the node discovery CRs
    - `Failed`: Failed to generate at least one node discovery CR
    - `Completed`: Successfully generated all the node discovery CRs```

Reply:
Phases must be part of API, i.e. this golang chunk.

Reply:
Converted it to enum